src/Aspire.Hosting/ResourceBuilderExtensions.cs

```diff
@@ -706,6 +709,123 @@ public static IResourceBuilder<T> WithHealthCheck<T>(this IResourceBuilder<T> bu
+            configure.AddFilter($"System.Net.Http.HttpClient.{healthCheckKey}.LogicalHandler", LogLevel.None);
+            configure.AddFilter($"System.Net.Http.HttpClient.{healthCheckKey}.ClientHandler", LogLevel.None);
+        });
+
+        builder.ApplicationBuilder.Services.AddHealthChecks().AddUrlGroup((UriHealthCheckOptions options) =>
+        {
+            if (uri is null)
+            {
+                throw new DistributedApplicationException($"The URI for the health check is not set. Ensure that the resource has been allocated before the health check is executed.");
+            }
+
+            options.AddUri(uri, setup => setup.ExpectHttpCode(statusCode ?? 200));
+        }, healthCheckKey);
+
+        builder.WithHealthCheck(healthCheckKey);
+
+        return builder;
+    }
+
+    /// <summary>
+    /// Adds a health check to the resource which is mapped to a specific endpoint.
+    /// </summary>
+    /// <typeparam name="T">A resource type that implements <see cref="IResourceWithEndpoints" />.</typeparam>
+    /// <param name="builder">A resource builder.</param>
+    /// <param name="path">The relative path to test.</param>
+    /// <param name="statusCode">The result code to interpret as healthy.</param>
+    /// <param name="endpointName">The name of the endpoint to derive the base address from.</param>
+    /// <returns>A resource builder.</returns>
+    /// <remarks>
+    /// <para>
+    /// This method adds a health check to the health check service which polls the specified endpoint on the resource
+    /// on a periodic basis. The base address is dynamically determined based on the endpoint that was selected. By
+    /// default the path is set to "/" and the status code is set to 200.
+    /// </para>
+    /// </remarks>
+    /// <example>
+    /// This example shows add a HTTPS health check to a backend project
+    /// to make sure that the front end does not start until the backend is
+    /// reporting a healthy status based on the return code returned from the
+    /// "/health" path on the backend server.
+    /// <code lang="C#">
+    /// var builder = DistributedApplication.CreateBuilder(args);
+    /// var backend = builder.AddProject&lt;Projects.Backend&gt;("backend")
+    ///                      .WithHttpsHealthCheck("/health");
+    /// builder.AddProject&lt;Projects.Backend&gt;("backend")
+    ///        .WithReference(backend).WaitFor(backend);
+    /// </code>
+    /// </example>
+    public static IResourceBuilder<T> WithHttpsHealthCheck<T>(this IResourceBuilder<T> builder, string? path = null, int? statusCode = null, string? endpointName = null) where T : IResourceWithEndpoints
+    {
+        endpointName = endpointName ?? "https";
+
+        if (builder.Resource.GetEndpoint(endpointName) is not { Scheme: "https" })
+        {
+            throw new DistributedApplicationException($"The endpoint '{endpointName}' on resource '{builder.Resource.Name}' was not using the 'https' scheme.");
+        }
```

Comment:
This should be deferred as well. Instead of having 2 checks we should just check for either http or https in the `WithHttpHealthCheck` method.